Spaceship app analytics

pajato1 opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
https://github.com/fastlane/fastlane/issues/12638

### Description
Adds App Analytics support to Spaceship.  For example:

  app = Spaceship::Tunes::Application.find("<your app>")
  app_analytics = app.analytics
  time = Time.now
  past = time - (60 * 60 * 24)
  end_t   = time.strftime("%Y-%m-%dT00:00:00Z")
  start_t = past.strftime("%Y-%m-%dT00:00:00Z")
  app_crashes = app_analytics.app_crashes_interval(start_t, end_t)
  pp app_crashes

pajato1 commented:
Run the bot again.

pajato1 commented:
Run the bot again.

pajato1 commented:
https://github.com/fastlane-old/spaceship/pull/190

pajato1 committed:
spaceship-app-analytics

joshdholtz reviewed (changes requested):
Thanks for reviving this feature! It looks like I'm getting some errors though 😬

```rb
app_analytics = app.analytics
puts "app_views: #{app_analytics.app_views}"
```

```bash
{"itcBaseUrl"=>"https://itunesconnect.apple.com", "message"=>"Make sure the URL is correct and try again.", "l10n"=>{"Crossfire-response-badrequest-title"=>"The URL is invalid.", "Crossfire-response-badrequest-message"=>"Make sure the URL is correct and try again."}, "errors"=>[{"message"=>"Invalid adam id: 1358899889", "messageTemplate"=>"Invalid adam id: 1358899889", "invalidValue"=>"[1358899889]"}]}
```

pajato1 commented:
@joshdholtz It's working for me, but maybe I'm initializing the app variable differently.  I'm doing this:

  Spaceship::Tunes.login(ENV['FASTLANE_USERNAME'], ENV['FASTLANE_PASSWORD'])
  app = Spaceship::Tunes::Application.find("com.tripadvisor.LocalPicks")
  app_analytics = app.analytics
  puts "app_views: #{app_analytics.app_views}"

Are you doing something different?

joshdholtz commented:
@pajato1 Fun fact! That app I was testing it on hasn't been released yet so it didn't have any stats 😉 Can you add a fix to handle that? 😇

Commit:
Added a user friendly error when analytics are not available due to the
app not being live.

pajato1 commented:
@joshdholtz I added the requested fix.

pajato1 commented:
@joshdholtz Can I get a code review, please?

joshdholtz commented on spaceship/lib/spaceship/tunes/application.rb:
```diff
@@ -136,6 +137,16 @@ def url
       end
 
+      def analytics
+        if self.latest_version.app_status != AppStatus::READY_FOR_SALE
```
I think this if statement might be too strong? Event if the latest version of the app isn't ready for sale, there could be analytics from a previous version that is currently live. Right? 🤔 

I _think_ we want to make sure `live_version` isn't nil?

Commit:
Made check less strong by checking that the live_version isn't nil
instead of checking that the latest status is READY_FOR_SALE.

pajato1 commented:
@joshdholtz I added the requested change.

joshdholtz commented on spaceship/lib/spaceship/tunes/tunes_client.rb:
```diff
@@ -510,6 +510,32 @@ def update_member_roles!(member, roles: [], apps: [])
+        req.body = data.to_json
+        req.headers['Content-Type'] = 'application/json'
+        req.headers['Referer'] = 'https://analytics.itunes.apple.com/'
+        req.headers['X-Requested-By'] = 'analytics.itunes.apple.com'
```
Can we delete this `Referer` and `X-Requested-By`? 🤔

joshdholtz commented on spaceship/spec/tunes/app_analytics_spec.rb:
```diff
@@ -0,0 +1,35 @@
+      analytics = app.analytics
+      units = analytics.app_units
+
+      pp(units)
```
I don't think this `pp` is needed? 🙃

joshdholtz commented on spaceship/spec/tunes/app_analytics_spec.rb:
```diff
@@ -0,0 +1,35 @@
+      TunesStubbing.itc_stub_analytics
+
+      analytics = app.analytics
+      units = analytics.app_units
```
Can we also add in calls (and `expects`) for all the other analytics methods that are supported so we can get a more complete code coverage? 😇

joshdholtz reviewed (changes requested):
A few more questions/comments then we should be 💯 

Can you also revert the changes to `fastlane/spec/fixtures/xcodeproj/automatic_code_signing.xcodeproj/project.pbxproj`? This was a side effect of a test (which was fixed a little bit ago)

Commit:
Responses to code review:

fastlane/spec/fixtures/xcodeproj/automatic_code_signing.xcodeproj/project.pbxproj

  * Reverted changes.

spaceship/lib/spaceship/tunes/tunes_client.rb

  * Removed Referer and X-Requested-By request headers.

spaceship/spec/tunes/app_analytics_spec.rb

  * Added code coverage for other analytics methods.

pajato1 commented:
@joshdholtz I added the requested changes.

Commit:
Fixed lint issues.

joshdholtz reviewed (changes requested):
Thanks running the linter and adding more tests! One last thing is it looks like the tests are failing

```sh
Real HTTP connections are disabled. Unregistered request: POST https://analytics.itunes.apple.com/analytics/api/v1/data/time-series with body '{"adamId":["898536088"],"dimensionFilters":[],"endTime":"2018-06-27T00:00:00Z","frequency":"DAY","group":null,"measures":["units"],"startTime":"2018-06-20T00:00:00Z"}' with headers {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3', 'Content-Type'=>'application/json', 'User-Agent'=>'Spaceship 2.96.1'}
```

This means the URL that the tests are trying to hit aren't matching up 100% with any of the stubbed requests which means its trying to hit the actual API 🙃

Commit:
Fixed analytics stubbing.

Commit:
Fixed lint issues.

pajato1 commented:
@joshdholtz The tests are passing again.  The problem was the stubs had hard coded dates, so the tests started failing the next day.  Rookie mistake.

joshdholtz reviewed (commented):
@pajato1 Hey! Have you tried running this like... super recently? I'm getting...

```
{"itcBaseUrl"=>"https://appstoreconnect.apple.com", "message"=>"Make sure the URL is correct and try again.", "l10n"=> "Crossfire-response-badrequest-message"=>"Make sure the URL is correct and try again."}, "errors"=>[]}
```

🤔 🤔 🤔

Commit:
It turns out this header is required.

pajato1 commented:
@joshdholtz It turns out the "X-Requested-By" header is required.  I've added it back in.

joshdholtz reviewed (approved):
We good! 👍
